ws/api/transport.go

```diff
@@ -0,0 +1,254 @@
+	"github.com/mainflux/mainflux"
+	log "github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	"github.com/mainflux/mainflux/ws"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	protocol = "ws"
+)
+
+var (
+	errUnauthorizedAccess = errors.New("missing or invalid credentials provided")
+	errMalformedSubtopic  = errors.New("malformed subtopic")
+)
+
+var (
+	upgrader = websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+		CheckOrigin:     func(r *http.Request) bool { return true },
+	}
+	auth              mainflux.ThingsServiceClient
+	logger            log.Logger
+	channelPartRegExp = regexp.MustCompile(`^/channels/([\w\-]+)/messages(/[^?]*)?(\?.*)?$`)
+)
+
+type subscription struct {
+	pubID    string
+	chanID   string
+	subtopic string
+	conn     *websocket.Conn
+	// client   *ws.Connclient
+	channel *ws.Channel
+}
+
+// MakeHandler returns http handler with handshake endpoint.
+func MakeHandler(svc ws.Service, tc mainflux.ThingsServiceClient, l log.Logger) http.Handler {
+	auth = tc
+	logger = l
+
+	mux := bone.New()
+	mux.GetFunc("/channels/:id/messages", handshake(svc))
+	mux.GetFunc("/channels/:id/messages/*", handshake(svc))
+	mux.GetFunc("/version", mainflux.Health(protocol))
+	mux.Handle("/metrics", promhttp.Handler())
+
+	return mux
+}
+
+func handshake(svc ws.Service) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		sub, err := authorize(r)
+		if err != nil {
+			switch err {
+			case errUnauthorizedAccess:
+				w.WriteHeader(http.StatusForbidden)
+				return
+			default:
+				logger.Warn(fmt.Sprintf("Failed to authorize: %s", err.Error()))
+				w.WriteHeader(http.StatusServiceUnavailable)
+				return
+			}
+		}
+
+		channelParts := channelPartRegExp.FindStringSubmatch(r.RequestURI)
+		if len(channelParts) < 2 {
+			logger.Warn("Empty channel id or malformed url")
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		sub.subtopic, err = parseSubTopic(channelParts[2])
+		if err != nil {
+			logger.Warn("Empty channel id or malformed url")
+			w.WriteHeader(http.StatusBadRequest)
+		}
+
+		// Create a new ws connection.
+		conn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Failed to upgrade connection to websocket: %s", err.Error()))
+			return
+		}
+
+		sub.conn = conn
+
+		c := ws.NewClient(conn, sub.pubID, logger)
+
+		logger.Debug(fmt.Sprintf("Successfully upgraded communication to WS on channel %s", sub.chanID))
+
+		// Subscribe to the channel
+		if err := svc.Subscribe(context.Background(), sub.pubID, sub.chanID, sub.subtopic, c); err != nil {
+			logger.Warn(fmt.Sprintf("Failed to subscribe to broker: %s", err.Error()))
+			if err == ws.ErrFailedConnection {
+				sub.conn.Close()
+				sub.channel.Closed <- true
+				return
+			}
+		}
+
+		sub.channel = ws.NewChannel()
+
+		wg := new(sync.WaitGroup)
+		wg.Add(2)
+
+		go sub.listen(wg)
+		go sub.broadcast(wg, svc)
```

Comment:
Decouple subscriber and service. You can use the channel for that (`broadcast` sends to it, and the service reads and does `Push`).

Reply:
subcription{} struct has `*websocket.Conn` field, which is needed to read and write messages using ws protocol. And holds `pubID`, `chanID` and `subtopic`, I need to use a struct to hold these, and they're all required in listen() and broadcast methods.